executor/adapter.go

```diff
@@ -1254,16 +1260,27 @@ func (a *ExecStmt) GetTextToLog() string {
 func (a *ExecStmt) observeStmtBeginForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionBegin(sqlDigest, planDigest)
+		a.stmtExecCtx = &stmtstats.StatementExecutionContext{
+			SQLDigest:  sqlDigest,
+			PlanDigest: planDigest,
+		}
+		vars.StmtStats.OnExecutionBegin(a.stmtExecCtx)
 		// This is a special logic prepared for TiKV's SQLExecCount.
 		vars.StmtCtx.KvExecCounter = vars.StmtStats.CreateKvExecCounter(sqlDigest, planDigest)
 	}
 }
 
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
-		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if a.stmtExecCtx == nil {
+			sqlDigest, planDigest := a.getSQLPlanDigest()
+			a.stmtExecCtx = &stmtstats.StatementExecutionContext{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+			}
+		}
+		a.stmtExecCtx.ExecDuration = time.Since(vars.StartTime) + vars.DurationParse
```

Comment:
This context modification is really wired. It should not be modified by the framework level.